File: backend/services/product_service.py
from fastapi import HTTPException
from sqlalchemy.orm import Session
from repositories.product_repository import ProductRepository, product_repo


class ProductService:

    def __init__(self, repo: ProductRepository = None):
        self.repo = repo or product_repo

    def get_all_products(self, db: Session, vendor_id=None, category_id=None):
        return self.repo.find_all(db, vendor_id=vendor_id, category_id=category_id)

    def get_product_by_id(self, db: Session, product_id: str):
        product = self.repo.find_by_id(db, product_id)
        if not product:
            raise HTTPException(status_code=404, detail="Produit introuvable")
        return product

    def create_product(self, db: Session, name: str, description, price, stock: int,
                       vendor_id: str, category_id=None):
        if float(price) <= 0:
            raise HTTPException(status_code=400, detail="Le prix doit être supérieur à 0")
        if stock < 0:
            raise HTTPException(status_code=400, detail="Le stock ne peut pas être négatif")
        return self.repo.create(db, name=name, description=description, price=price,
                                stock=stock, vendor_id=vendor_id, category_id=category_id)

    def update_product(self, db: Session, product_id: str, data: dict,
                       vendor_id: str, user_role: str):
        product = self.repo.find_by_id(db, product_id)
        if not product:
            raise HTTPException(status_code=404, detail="Produit introuvable")
        if user_role != "admin" and product["vendor_id"] != vendor_id:
            raise HTTPException(status_code=403, detail="Non autorisé à modifier ce produit")
        return self.repo.update(db, product_id, **{k: v for k, v in data.items() if v is not None})

    def delete_product(self, db: Session, product_id: str, vendor_id: str, user_role: str):
        product = self.repo.find_by_id(db, product_id)
        if not product:
            raise HTTPException(status_code=404, detail="Produit introuvable")
        if user_role != "admin" and product["vendor_id"] != vendor_id:
            raise HTTPException(status_code=403, detail="Non autorisé à supprimer ce produit")
        self.repo.delete(db, product_id)

    def get_categories(self, db: Session):
        return self.repo.find_categories(db)


product_service = ProductService()

File: backend/src/services/product.service.js
const productRepository = require('../repositories/product.repository');

class ProductService {
  async getAllProducts(filters = {}) {
    return productRepository.findAll(filters);
  }

  async getProductById(id) {
    const product = await productRepository.findById(id);
    if (!product) {
      const err = new Error('Produit introuvable');
      err.status = 404;
      throw err;
    }
    return product;
  }

  async createProduct({ name, description, price, stock, categoryId }, vendorId) {
    if (price <= 0) {
      const err = new Error('Le prix doit être supérieur à 0');
      err.status = 400;
      throw err;
    }
    if (stock < 0) {
      const err = new Error('Le stock ne peut pas être négatif');
      err.status = 400;
      throw err;
    }
    return productRepository.create({ name, description, price, stock, vendorId, categoryId });
  }

  async updateProduct(id, data, vendorId, userRole) {
    const product = await productRepository.findById(id);
    if (!product) {
      const err = new Error('Produit introuvable');
      err.status = 404;
      throw err;
    }
    // Seul le vendeur propriétaire ou un admin peut modifier
    if (userRole !== 'admin' && product.vendor_id !== vendorId) {
      const err = new Error('Non autorisé à modifier ce produit');
      err.status = 403;
      throw err;
    }
    return productRepository.update(id, data);
  }

  async deleteProduct(id, vendorId, userRole) {
    const product = await productRepository.findById(id);
    if (!product) {
      const err = new Error('Produit introuvable');
      err.status = 404;
      throw err;
    }
    if (userRole !== 'admin' && product.vendor_id !== vendorId) {
      const err = new Error('Non autorisé à supprimer ce produit');
      err.status = 403;
      throw err;
    }
    await productRepository.delete(id);
  }

  async getCategories() {
    return productRepository.findCategories();
  }
}

module.exports = new ProductService();
